cmd/kube-proxy/app/server.go

```diff
@@ -88,10 +90,28 @@ func (s *ProxyServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.HostnameOverride, "hostname-override", s.HostnameOverride, "If non-empty, will use this string as identification instead of the actual hostname.")
 	fs.BoolVar(&s.ForceUserspaceProxy, "legacy-userspace-proxy", true, "Use the legacy userspace proxy (instead of the pure iptables proxy).")
 	fs.DurationVar(&s.SyncPeriod, "iptables-sync-period", 5*time.Second, "How often iptables rules are refreshed (e.g. '5s', '1m', '2h22m').  Must be greater than 0.")
+	fs.BoolVar(&s.CleanupAndExit, "cleanup-iptables", false, "If true cleanup iptables rules and exit.")
 }
 
-// Run runs the specified ProxyServer.  This should never exit.
+// Run runs the specified ProxyServer.  This should never exit (unless CleanupAndExit is set).
 func (s *ProxyServer) Run(_ []string) error {
+	protocol := utiliptables.ProtocolIpv4
+	if s.BindAddress.To4() == nil {
+		protocol = utiliptables.ProtocolIpv6
+	}
+
+	// remove iptables rules and exit
+	if s.CleanupAndExit {
+		execer := exec.New()
+		ipt := utiliptables.New(execer, protocol)
+		encounteredError := userspace.TearDownIptablesProxierRules(ipt)
```

Comment:
That sounds nice. Should have it up in a sec.
On Aug 20, 2015 6:57 PM, "Tim Hockin" <notifications@github.com> wrote:

> In cmd/kube-proxy/app/server.go
> <https://github.com/kubernetes/kubernetes/pull/12981#discussion_r37590663>
> :
>
> >  func (s *ProxyServer) Run(_ []string) error {
> > +	protocol := utiliptables.ProtocolIpv4
> > +	if s.BindAddress.To4() == nil {
> > +		protocol = utiliptables.ProtocolIpv6
> > +	}
> > +
> > +	// remove iptables rules and exit
> > +	if s.CleanupAndExit {
> > +		execer := exec.New()
> > +		ipt := utiliptables.New(execer, protocol)
> > +		encounteredError := userspace.TearDownIptablesProxierRules(ipt)
>
> Funnily, these are now ass-backwards. Fixing it will be a net improvement.
> Move the iptables cleanup into the iptables package and the userspace
> cleanup into the userspace package, and have them each call each other.
>
> Edit: damn, that's a circular dep. The problem with this mode is that when
> I delete the userspace module, I will have to change this code, which is
> just not the right thing.
>
> How about this: Move each proxier's cleanup into the correct proxier pkg,
> make then exported symbols. In cmd/kube-proxy/.. call the other proxier's
> cleanup after NewProxy() returns non-nil.
>
> e.g.
>
> if useIptables {
>     p := iptables.NewProxier(ipt)
>     if p != nil {
>         userspace.CleanupLeftovers()
>     }
> else
>     .... opposite
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12981/files#r37590663>.
>